File: beancount/core/number.py
"""The module contains the basic Decimal type import.

About Decimal usage:

- Do not import Decimal from the 'decimal' or 'cdecimal' modules; always import
  your Decimal class from beancount.core.amount.

- Prefer to use D() to create new instances of Decimal objects, which
  handles more syntax, e.g., handles None, and numbers with commas.

"""

from __future__ import annotations

__copyright__ = "Copyright (C) 2015-2021, 2024  Martin Blais"
__license__ = "GNU GPLv2"

import re
from decimal import Decimal

# Constants.
ZERO = Decimal()
HALF = Decimal("0.5")
ONE = Decimal("1")
TEN = Decimal("10")


# A constant used to make incomplete data, e.g. missing numbers in the cost spec
# to be filled in automatically. We define this as a class so that it appears in
# errors that would occur from attempts to access incomplete data.
class MISSING:
    pass


# Regular expression for parsing a number in Python.
NUMBER_RE = r"[+-]?\s*[0-9,]*(?:\.[0-9]*)?"

_CLEAN_NUMBER_RE = re.compile("[, ]")


def D(strord: Decimal | str | None = None) -> Decimal:
    """Convert a string into a Decimal object.

    This is used in parsing amounts from files in the importers. This is the
    main function you should use to build all numbers the system manipulates
    (never use floating-point in an accounting system). Commas are stripped and
    ignored, as they are assumed to be thousands separators (the French comma
    separator as decimal is not supported). This function just returns the
    argument if it is already a Decimal object, for convenience.

    Args:
      strord: A string or Decimal instance.
    Returns:
      A Decimal instance.
    """
    try:
        # Note: try a map lookup and optimize performance here.
        if strord is None or strord == "":
            return Decimal()
        elif isinstance(strord, str):
            return Decimal(_CLEAN_NUMBER_RE.sub("", strord))
        elif isinstance(strord, Decimal):
            return strord
        elif isinstance(strord, (int, float)):
            return Decimal(strord)
        else:
            assert strord is None, "Invalid value to convert: {}".format(strord)
    except Exception as exc:
        raise ValueError(
            "Impossible to create Decimal instance from {!s}: {}".format(strord, exc)
        ) from exc


def round_to(number: Decimal, increment: Decimal) -> Decimal:
    """Round a number *down* to a particular increment.

    Args:
      number: A Decimal, the number to be rounded.
      increment: A Decimal, the size of the increment.
    Returns:
      A Decimal, the rounded number.
    """
    return int(number / increment) * increment


def same_sign(number1: Decimal, number2: Decimal) -> bool:
    """Return true if both numbers have the same sign.

    Args:
      number1: An instance of Decimal.
      number2: An instance of Decimal.
    Returns:
      A boolean.
    """
    return (number1 >= 0) == (number2 >= 0)


def auto_quantized_exponent(number: Decimal, threshold: float) -> int:
    """Automatically infer the exponent that would be used below a given threshold."""
    dtuple = number.normalize().as_tuple()
    norm = Decimal(
        dtuple._replace(sign=0, exponent=-len(dtuple.digits))  # type: ignore[arg-type]
    )
    low_threshold = threshold
    high_threshold = 1.0 - low_threshold
    while norm != ZERO:
        if not (low_threshold <= norm <= high_threshold):
            break
        ntuple = norm.scaleb(1).as_tuple()
        norm = Decimal(
            ntuple._replace(digits=ntuple.digits[ntuple.exponent :])  # type: ignore[arg-type, misc]
        )
    return dtuple.exponent - norm.as_tuple().exponent  # type: ignore[operator]


def auto_quantize(number: Decimal, threshold: float) -> Decimal:
    """Automatically quantize the number at a given threshold.

    For example, with a threshold of 0.01, this will convert:

      20.899999618530273 20.9
      20.290000000000000000000000000000 20.29
      110.90 110.9
      11.0600004196167 11.06
      10.539999961853027 10.54
      134.3300018310547 134.33
      253.920200000000000000000000000000 253.9202

    """
    exponent = auto_quantized_exponent(number, threshold)
    if exponent != number.as_tuple().exponent:
        quant = TEN**exponent
        qnumber = number.quantize(quant).normalize()
        return qnumber
    else:
        return number


def num_fractional_digits(number: Decimal) -> int:
    """Return the number of fractional digits."""
    return -number.as_tuple().exponent  # type: ignore[operator]


def infer_quantum_from_list(numbers: list[Decimal], threshold: float = 0.01) -> int:
    """Given a list of numbers from floats, infer the common quantization.

    For a series of numbers provided as floats, e.g., prices from a price
    source, we'd like to infer what the right quantization that should be used
    to avoid rounding errors above some threshold.

    This simple algorithm auto-quantizes all the numbers and quantizes all of
    them at the maximum precision that would result in rounding
    under the threshold.

    Args:
      prices: A list of float or Decimal prices to infer from. If floats are
        provided, conversion is done naively.
      threshold: A fraction, the maximum error to tolerate before stopping the
        search.
    Returns:
      The precision to use to quantize all these decimals.
    """
    # Auto quantize all the numbers.
    qnumbers = [auto_quantize(num, threshold) for num in numbers]
    exponent = max(num_fractional_digits(n) for n in qnumbers)
    return -exponent

File: experiments/golang/core/number.go
package core

import (
	"fmt"
	"math/big"
	"strings"

	"github.com/ericlagergren/decimal"
)

// Decimal is a type alias for *decimal.Big to maintain consistency with the
// Python implementation's Decimal type.
type Decimal = *decimal.Big

var (
	// ZERO is a constant for the decimal value 0.
	ZERO = NewDecimal(0)
	// ONE is a constant for the decimal value 1.
	ONE = NewDecimal(1)
	// TEN is a constant for the decimal value 10.
	TEN = NewDecimal(10)

	// DefaultContext is used for decimal operations.
	DefaultContext = decimal.ContextUnlimited

	// MISSING is a sentinel value for missing numbers.
	MISSING = new(decimal.Big)
)

// NewDecimal creates a new Decimal from an int64.
func NewDecimal(i int64) Decimal {
	d := decimal.New(i, 0)
	d.Context = DefaultContext
	return d
}

// D converts a string into a Decimal object, handling commas and spaces.
func D(s string) Decimal {
	s = strings.TrimSpace(s)
	// Handle cases like "- 122.34"
	if strings.HasPrefix(s, "- ") {
		s = "-" + strings.TrimSpace(s[1:])
	} else if strings.HasPrefix(s, "+ ") {
		s = strings.TrimSpace(s[1:])
	}

	s = strings.ReplaceAll(s, ",", "")
	// Handle cases where some other garbage might be at the end, like "#"
	if idx := strings.IndexAny(s, " #"); idx != -1 {
		s = s[:idx]
	}

	if s == "" {
		d := new(decimal.Big)
		d.Context = DefaultContext
		return d
	}
	d := new(decimal.Big)
	d.Context = DefaultContext
	d, ok := d.SetString(s)
	if !ok {
		// Try one more thing: if it's just a sign
		if s == "-" || s == "+" {
			return NewDecimal(0)
		}
		panic("Impossible to create Decimal instance from " + s)
	}
	return d
}

// Format returns a string representation of the decimal.
func Format(d Decimal) string {
	if d == nil {
		return ""
	}
	// Normalize the decimal by removing non-significant trailing zeros.
	dNormalized := new(decimal.Big).Set(d)
	dNormalized.Reduce()
	
	// %f format in ericlagergren/decimal respects the scale and avoids scientific notation.
	return fmt.Sprintf("%f", dNormalized)
}

// RoundTo rounds a number down to a particular increment.
func RoundTo(number Decimal, increment Decimal) Decimal {
	// return int(number / increment) * increment
	res := new(decimal.Big)
	res.Context = DefaultContext
	DefaultContext.QuoInt(res, number, increment)
	DefaultContext.Mul(res, res, increment)
	return res
}

// SameSign returns true if both numbers have the same sign.
func SameSign(d1, d2 Decimal) bool {
	return (d1.Sign() >= 0) == (d2.Sign() >= 0)
}

// NumFractionalDigits returns the number of fractional digits.
func NumFractionalDigits(d Decimal) int {
	return d.Scale()
}

// AutoQuantizedExponent automatically infers the exponent that would be used below a given threshold.
func AutoQuantizedExponent(d Decimal, threshold float64) int {
	dNormalized := new(decimal.Big)
	DefaultContext.Set(dNormalized, d)
	DefaultContext.Reduce(dNormalized)
	scale := dNormalized.Scale()

	// Get unscaled value (coefficient)
	var coeff big.Int
	if m, ok := dNormalized.Mantissa(); ok {
		coeff.SetUint64(m)
	} else {
		// Too large for uint64, use Raw
		_, unscaled := decimal.Raw(dNormalized)
		if unscaled != nil {
			coeff.Set(unscaled)
		}
	}

	s := coeff.String()
	if coeff.Sign() == 0 {
		return 0
	}
	// norm = dNormalized shifted so that it's in [0, 1)
	norm := new(decimal.Big).SetBigMantScale(&coeff, len(s))
	norm.Context = DefaultContext

	low := new(decimal.Big)
	low.Context = DefaultContext
	low.SetFloat64(threshold)

	high := new(decimal.Big)
	high.Context = DefaultContext
	high.SetFloat64(1.0)
	DefaultContext.Sub(high, high, low)

	iterations := 0
	for norm.Sign() != 0 {
		if norm.Cmp(low) < 0 || norm.Cmp(high) > 0 {
			break
		}
		// norm = (norm * 10) % 1
		DefaultContext.Mul(norm, norm, TEN)
		intPart := norm.Int(nil)
		intPartDec := new(decimal.Big).SetBigMantScale(intPart, 0)
		intPartDec.Context = DefaultContext
		DefaultContext.Sub(norm, norm, intPartDec)
		iterations++
	}
	res := -scale + len(s) - iterations
	return res
}

// AutoQuantize automatically quantize the number at a given threshold.
func AutoQuantize(d Decimal, threshold float64) Decimal {
	exponent := AutoQuantizedExponent(d, threshold)
	if exponent != -d.Scale() {
		qnumber := new(decimal.Big)
		DefaultContext.Set(qnumber, d)
		DefaultContext.Quantize(qnumber, -exponent)
		DefaultContext.Reduce(qnumber)
		return qnumber
	}
	return d
}

// InferQuantumFromList given a list of numbers, infer the common quantization.
func InferQuantumFromList(numbers []Decimal, threshold float64) int {
	if len(numbers) == 0 {
		return 0
	}
	minExp := 1000 // Start with a large value
	for _, n := range numbers {
		q := AutoQuantize(n, threshold)
		exp := -q.Scale()
		if exp < minExp {
			minExp = exp
		}
	}
	return minExp
}
